pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/groupby/DefaultGroupByExecutor.java

```diff
@@ -271,15 +238,70 @@ public AggregationGroupByResult getResult() {
    * For SV keys: _docIdToSVGroupKey mapping is updated.
    * For MV keys: _docIdToMVGroupKey mapping is updated.
    *
-   * @param docIdSet
-   * @param startIndex
-   * @param length
+   * @param projectionBlock Projection block for which to generate group keys
    */
-  private void generateGroupKeysForDocIdSet(int[] docIdSet, int startIndex, int length) {
+  private void generateGroupKeysForBlock(ProjectionBlock projectionBlock) {
     if (_hasMultiValuedColumns) {
-      _groupKeyGenerator.generateKeysForDocIdSet(docIdSet, startIndex, length, _docIdToMVGroupKey);
+      _groupKeyGenerator.generateKeysForBlock(projectionBlock, _docIdToMVGroupKey);
     } else {
-      _groupKeyGenerator.generateKeysForDocIdSet(docIdSet, startIndex, length, _docIdToSVGroupKey);
+      _groupKeyGenerator.generateKeysForBlock(projectionBlock, _docIdToSVGroupKey);
+    }
+  }
+
+  /**
+   * Helper method to initialize result holder array.
+   *
+   * @param aggregationInfoList List of aggregation infos.
+   * @param trimSize Trim size for group by keys
+   * @param maxNumResults Maximum number of groups possible
+   */
+  private void initResultHolderArray(List<AggregationInfo> aggregationInfoList, int trimSize, int maxNumResults) {
+    _aggrFuncContextArray = new AggregationFunctionContext[_numAggrFunc];
+    _resultHolderArray = new GroupByResultHolder[_numAggrFunc];
+
+    for (int i = 0; i < _numAggrFunc; i++) {
+      AggregationInfo aggregationInfo = aggregationInfoList.get(i);
+      String[] columns = aggregationInfo.getAggregationParams().get("column").trim().split(",");
+
+      _aggrFuncContextArray[i] = new AggregationFunctionContext(aggregationInfo.getAggregationType(), columns);
+      _resultHolderArray[i] =
+          ResultHolderFactory.getGroupByResultHolder(_aggrFuncContextArray[i].getAggregationFunction(), maxNumResults,
+              trimSize);
     }
   }
+
+  /**
+   * Allocate storage for docId to group keys mapping.
+   */
+  private void initDocIdToGroupKeyMap() {
+    if (_hasMultiValuedColumns) {
+      _docIdToMVGroupKey = new int[DocIdSetPlanNode.MAX_DOC_PER_CALL][];
+    } else {
+      _docIdToSVGroupKey = new int[DocIdSetPlanNode.MAX_DOC_PER_CALL];
+    }
+  }
+
+  /**
+   * Initializes the following:
+   * <p> - Group key generator. </p>
+   * <p> - Result holders </p>
+   * <p> - Re-usable storage (eg docId to group key mapping) </p>
+   *
+   * This is separate from init(), as this can only happen within process as projection block is
+   * required to create group key generator.
+   *
+   * @param projectionBlock Projection block to group by.
+   */
+  private void initGroupBy(ProjectionBlock projectionBlock) {
+    if (_groupByInited) {
```

Comment:
I prefer this way in cases I don't want caller to worry about keeping track of whether initialization has happened or not.